File: ask.py
import os
import time
import argparse
from dotenv import load_dotenv
from openai import OpenAI
from anthropic import Anthropic
from tabulate import tabulate

from cache import get_cached_response, save_cached_response

load_dotenv()

openai_client = OpenAI(api_key=os.getenv("OPENAI_API_KEY"))
anthropic_client = Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

OPENAI_MODEL = "gpt-4.1-mini"
CLAUDE_MODEL = "claude-haiku-4-5"


def ask_gpt(prompt, system):
    cached = get_cached_response("gpt", prompt, system)

    if cached:
        cached["cached"] = True
        return cached

    start = time.time()

    response = openai_client.responses.create(
        model=OPENAI_MODEL,
        input=[
            {"role": "system", "content": system},
            {"role": "user", "content": prompt},
        ],
    )

    latency = time.time() - start

    result = {
        "model": OPENAI_MODEL,
        "answer": response.output_text,
        "input_tokens": response.usage.input_tokens,
        "output_tokens": response.usage.output_tokens,
        "latency": round(latency, 2),
        "cached": False,
    }

    save_cached_response("gpt", prompt, system, result)

    return result


def ask_claude(prompt, system):
    cached = get_cached_response("claude", prompt, system)

    if cached:
        cached["cached"] = True
        return cached

    start = time.time()

    response = anthropic_client.messages.create(
        model=CLAUDE_MODEL,
        max_tokens=500,
        system=system,
        messages=[
            {"role": "user", "content": prompt},
        ],
    )

    latency = time.time() - start

    result = {
        "model": CLAUDE_MODEL,
        "answer": response.content[0].text,
        "input_tokens": response.usage.input_tokens,
        "output_tokens": response.usage.output_tokens,
        "latency": round(latency, 2),
        "cached": False,
    }

    save_cached_response("claude", prompt, system, result)

    return result


def stream_gpt(prompt, system):
    print("\n===== GPT STREAM =====\n")

    stream = openai_client.responses.create(
        model=OPENAI_MODEL,
        input=[
            {"role": "system", "content": system},
            {"role": "user", "content": prompt},
        ],
        stream=True,
    )

    for event in stream:
        if event.type == "response.output_text.delta":
            print(event.delta, end="", flush=True)

    print("\n")


def stream_claude(prompt, system):
    print("\n===== CLAUDE STREAM =====\n")

    with anthropic_client.messages.stream(
        model=CLAUDE_MODEL,
        max_tokens=500,
        system=system,
        messages=[
            {"role": "user", "content": prompt},
        ],
    ) as stream:
        for text in stream.text_stream:
            print(text, end="", flush=True)

    print("\n")


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("prompt")
    parser.add_argument(
        "--system",
        default="You are a helpful assistant.",
    )
    parser.add_argument(
        "--stream",
        action="store_true",
    )

    args = parser.parse_args()

    if args.stream:
        stream_gpt(args.prompt, args.system)
        stream_claude(args.prompt, args.system)
        return

    gpt_result = ask_gpt(args.prompt, args.system)
    claude_result = ask_claude(args.prompt, args.system)

    table = [
        [
            "GPT",
            gpt_result["answer"],
            gpt_result["input_tokens"],
            gpt_result["output_tokens"],
            gpt_result["latency"],
            gpt_result["cached"],
        ],
        [
            "Claude",
            claude_result["answer"],
            claude_result["input_tokens"],
            claude_result["output_tokens"],
            claude_result["latency"],
            claude_result["cached"],
        ],
    ]

    print(
        tabulate(
            table,
            headers=[
                "Provider",
                "Answer",
                "Input Tokens",
                "Output Tokens",
                "Latency",
                "Cached",
            ],
            tablefmt="grid",
        )
    )


if __name__ == "__main__":
    main()

File: ask.js
import "dotenv/config";
import OpenAI from "openai";
import Anthropic from "@anthropic-ai/sdk";

import {
  getCachedResponse,
  saveCachedResponse,
} from "./cache.js";

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

const anthropic = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY,
});

const prompt = process.argv[2];
const isStreaming = process.argv.includes("--stream");

const systemPrompt = "You are a helpful assistant.";

if (!prompt) {
  console.log('Usage: node ask.js "your prompt" --stream');
  process.exit(1);
}

async function askGPT() {
  const cached = await getCachedResponse("gpt", prompt, systemPrompt);

  if (cached) {
    console.log("Using cached GPT response");
    return cached;
  }

  const start = Date.now();

  const response = await openai.responses.create({
    model: "gpt-4.1-mini",
    input: [
      { role: "system", content: systemPrompt },
      { role: "user", content: prompt },
    ],
  });

  const result = {
    provider: "GPT",
    model: "gpt-4.1-mini",
    answer: response.output_text,
    input_tokens: response.usage.input_tokens,
    output_tokens: response.usage.output_tokens,
    latency: Number(((Date.now() - start) / 1000).toFixed(2)),
  };

  await saveCachedResponse("gpt", prompt, systemPrompt, result);

  return result;
}

async function askClaude() {
  const cached = await getCachedResponse("claude", prompt, systemPrompt);

  if (cached) {
    console.log("Using cached Claude response");
    return cached;
  }

  const start = Date.now();

  const response = await anthropic.messages.create({
    model: "claude-haiku-4-5",
    max_tokens: 500,
    system: systemPrompt,
    messages: [{ role: "user", content: prompt }],
  });

  const result = {
    provider: "Claude",
    model: "claude-haiku-4-5",
    answer: response.content[0].text,
    input_tokens: response.usage.input_tokens,
    output_tokens: response.usage.output_tokens,
    latency: Number(((Date.now() - start) / 1000).toFixed(2)),
  };

  await saveCachedResponse("claude", prompt, systemPrompt, result);

  return result;
}

async function streamGPT() {
  console.log("\n===== GPT STREAM =====\n");

  const response = await openai.responses.create({
    model: "gpt-4.1-mini",
    input: [
      { role: "system", content: systemPrompt },
      { role: "user", content: prompt },
    ],
    stream: true,
  });

  for await (const event of response) {
    if (event.type === "response.output_text.delta") {
      process.stdout.write(event.delta);
    }
  }

  console.log("\n");
}

async function streamClaude() {
  console.log("\n===== CLAUDE STREAM =====\n");

  const stream = anthropic.messages.stream({
    model: "claude-haiku-4-5",
    max_tokens: 500,
    system: systemPrompt,
    messages: [{ role: "user", content: prompt }],
  });

  for await (const event of stream) {
    if (
      event.type === "content_block_delta" &&
      event.delta &&
      event.delta.text
    ) {
      process.stdout.write(event.delta.text);
    }
  }

  console.log("\n");
}

async function main() {
  try {
    if (isStreaming) {
      await streamGPT();
      await streamClaude();
      return;
    }

    const gpt = await askGPT();
    const claude = await askClaude();
    
console.log("\n================ GPT RESULT ================");
console.log(`Model: ${gpt.model}`);
console.log(`Input Tokens: ${gpt.input_tokens}`);
console.log(`Output Tokens: ${gpt.output_tokens}`);
console.log(`Latency: ${gpt.latency}s`);
console.log("\nAnswer:");
console.log(gpt.answer);

console.log("\n================ CLAUDE RESULT ================");
console.log(`Model: ${claude.model}`);
console.log(`Input Tokens: ${claude.input_tokens}`);
console.log(`Output Tokens: ${claude.output_tokens}`);
console.log(`Latency: ${claude.latency}s`);
console.log("\nAnswer:");
console.log(claude.answer);
  } catch (error) {
    console.error("Error:", error.message);
  }
}

main();